internal/resources/free_tier_cluster.go

```diff
@@ -0,0 +1,582 @@
+	}
+	plan.AppServiceId = types.StringNull()
+	plan.ConnectionString = types.StringNull()
+	plan.Audit = types.ObjectNull(providerschema.CouchbaseAuditData{}.AttributeTypes())
+	plan.Availability = types.ObjectNull(providerschema.Availability{}.AttributeTypes())
+	plan.CmekId = types.StringNull()
+
+	if plan.ServiceGroups.IsNull() || plan.ServiceGroups.IsUnknown() {
+		plan.ServiceGroups = types.SetNull(types.ObjectType{}.WithAttributeTypes(providerschema.ServiceGroupAttributeTypes()))
+	}
+	return plan
+}
+
+// checkFreeTierClusterStatus monitors the status of a cluster creation, update and deletion operation for a specified
+// organization, project, and cluster ID. It periodically fetches the cluster status using the `getCluster`
+// function and waits until the cluster reaches a final state or until a specified timeout is reached.
+// The function returns an error if the operation times out or encounters an error during status retrieval.
+func (c *FreeTierCluster) checkFreeTierClusterStatus(ctx context.Context, organizationId, projectId, ClusterId string) error {
+	var (
+		clusterResp *freeTierClusterapi.GetFreeTierClusterResponse
+		err         error
+	)
+
+	// Assuming 60 minutes is the max time deployment takes, can change after discussion
+	const timeout = time.Minute * 60
+
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	const sleep = time.Second * 3
+
+	timer := time.NewTimer(2 * time.Minute)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("cluster creation status transition timed out after initiation, unexpected error: %w", err)
+		case <-timer.C:
+			clusterResp, err = c.getFreeTierCluster(ctx, organizationId, projectId, ClusterId)
+			switch err {
+			case nil:
+				if clusterapi.IsFinalState(clusterapi.State(clusterResp.CurrentState)) {
+					return nil
+				}
+				const msg = "waiting for cluster to complete the execution"
+				tflog.Info(ctx, msg)
+			default:
+				return err
+			}
+			timer.Reset(sleep)
+		}
+	}
+}
+
+// getCluster retrieves cluster information from the specified organization and project
+// using the provided cluster ID by open-api call.
+func (f *FreeTierCluster) getFreeTierCluster(ctx context.Context, organizationId, projectId, clusterId string) (*freeTierClusterapi.GetFreeTierClusterResponse, error) {
+	url := fmt.Sprintf("%s/v4/organizations/%s/projects/%s/clusters/freeTier/%s", f.HostURL, organizationId, projectId, clusterId)
+	cfg := api.EndpointCfg{Url: url, Method: http.MethodGet, SuccessStatus: http.StatusOK}
+	response, err := f.Client.ExecuteWithRetry(
+		ctx,
+		cfg,
+		nil,
+		f.Token,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", errors.ErrExecutingRequest, err)
+	}
+
+	clusterResp := freeTierClusterapi.GetFreeTierClusterResponse{}
+	err = json.Unmarshal(response.Body, &clusterResp)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", errors.ErrUnmarshallingResponse, err)
```

Comment:
define errors in this file:
https://github.com/couchbasecloud/terraform-provider-couchbase-capella/blob/ca8a41bf6f11dd49c30d941f9b6cb3bec3d6bae5/internal/errors/errors.go

then use the value here